pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,329 @@
+    if (!IsProcessRunning(controller_pid)) {
+      return;
+    }
+    timer_->SleepUs(10 * 1000 /* 0.01s */);
+  }
+  handler_->Message(
+      kWarning, "Couldn't get controller process (%d) to exit", controller_pid);
+}
+
+void ControllerManager::RunController() {
+  handler_->Message(kInfo, "Controller running with PID %d", getpid());
+
+  // Would set up gRPC server here and pass control to its event loop.  Instead
+  // just hang out sleeping forever.
+  while (true) {
+    sleep(60);  // 1m
+  }
+}
+
+void ControllerManager::Daemonize() {
+  // Make a new session (process group).
+  if (setsid() < 0) {
+    handler_->Message(kWarning, "Daemonize: Failed to setsid().");
+  }
+
+  // We need to fork again to make sure there is no session group leader.
+  pid_t pid = fork();
+  CHECK(pid != -1) << "Couldn't fork to daemonize.";
+  if (pid != 0) {
+    exit(EXIT_SUCCESS);
+  }
+
+  // If we keep the current directory we might keep them from being able to
+  // unmount their filesystem.
+  if (chdir("/") < 0) {
+    handler_->Message(kWarning, "Daemonize: Failed to chdir(/).");
+  }
+
+  // If we disconnect file descriptors then logging will break, so don't.
+}
+
+void ControllerManager::ForkOffControllerProcess() {
+  handler_->Message(kInfo, "Forking controller process off of %d", getpid());
+
+  pid_t pid = fork();
+  CHECK(pid != -1) << "Couldn't fork a controller babysitter process";
+
+  if (pid != 0) return;  // Parent process.
+
+  // Now we're in the child process.  Set this up as a babysitter process,
+  // that forks off a controller and restarts it if it dies.
+
+  Daemonize();
+  SetUpSignalHandlers();
+
+  factory_->set_is_root_process(false);
+  factory_->PrepareForkedProcess("babysitter");
+
+  babysitter_pid_ = getpid();
+  handler_->Message(kInfo, "Babysitter running with PID %d", babysitter_pid_);
+
+  KillControllerIfAlreadyRunning();
+  WriteControllerPid();
+  pid_t master_pid = MasterPid();  // Blocks until master process is seen.
+
+  while (true) {
+    pid = fork();
+    CHECK(pid != -1) << "Couldn't fork a controller process";
+
+    if (pid == 0) {
+      factory_->PrepareForkedProcess("controller");
+      factory_->PrepareControllerProcess();
+
+      // Start a thread to watch to see if the master or babysitter process
+      // dies, and quit if it does.
+      process_death_watcher_thread_.reset(
+          new ProcessDeathWatcherThread(thread_system_,
+                                        timer_,
+                                        handler_,
+                                        master_pid,
+                                        babysitter_pid_));
+      CHECK(process_death_watcher_thread_->Start());
+
+      RunController();
+    } else {
+      // Wait for controller process to die, then continue with the loop by
+      // restarting it.
+      int status;
+      pid_t child_pid = waitpid(pid, &status, 0);
+      CHECK(child_pid != -1) << "Call to waitpid failed.";
```

Comment:
http://pubs.opengroup.org/onlinepubs/7908799/xsh/wait.html suggests EINTR is a possibility, even if Linux's version only seems to issue it in weird corner case...